### ISSUE TITLE
lunr.Index.invertedIndex grows exponentially in size after each query

### ISSUE THREAD

**Issue by i80and** [olivernn/lunr.js#282]
I'm using lunr 2.1.0 with the following plugin:

```python
    function tokenPositionPlugin(builder) {
        let position = 0
        var pipelineFunction = function (token) {
            position += 1
            token.metadata['pos'] = position
            return token
        }
        lunr.Pipeline.registerFunction(pipelineFunction, 'tokenPositionMetadata')
        builder.pipeline.before(lunr.stemmer, pipelineFunction)
        builder.metadataWhitelist.push('pos')
    }
```

I'm using this positional data to implement phrase search. All is well and fine, except that `lunr.Index.intertedIndex[term][field]['pos']` grows in size after each query (by a factor of four in my testing, but that might be query dependent; I'm not sure).

This appears to be because ``MatchData.combine`` is mutating the InvertedIndex metadata, rather than a copy. The solution that seems to be working for me is to take a shallow copy of the `InvertedIndex[term][field]` object during `lunr.Index.query()`, as seen in https://github.com/i80and/lunr.js/commit/41c7db350c269b3df64543321df04307720fc617.

**Comment by olivernn** [comment]
That doesn't sound good! I'll have a look at re-creating this and then providing a patch.

Also, as an aside, pipeline functions are called with three arguments, one of which is the index of the token within the document/field being searched, e.g. you can just do the following in your pipeline function:

```javascript
var pipelineFunction = function (token, pos) {
  token.metadata['pos'] = pos
  return token
}
```

This is [documented](https://github.com/olivernn/lunr.js/blob/master/lib/pipeline.js#L41-L60), but it looks like JSDoc doesn't like how.

**Comment by i80and** [comment]
I can put together a minimal proof if you have trouble reproducing this issue.

And thanks for the advice! :D I really do love this library; I've been using it for a number of projects over the past couple of years, and it's always delivered.

**Comment by olivernn** [comment]
If you wouldn't mind trying to put together a reproduction, I tried to re-create it [here](https://jsfiddle.net/fkz51use/3/) but not getting the behaviour you mention.

I can see where in the code this would be possible, I just can't trigger it at the moment.

**Comment by i80and** [comment]
It looks like it only triggers when using multiple identical terms in a query. I'm using this for fuzzy-searching as you described here: https://github.com/olivernn/lunr.js/issues/276#issuecomment-308205156

https://github.com/i80and/lunr-bug-reproduction

```
16
32
64
128
256
512
1024
2048
4096
8192
16384
32768
65536

<--- Last few GCs --->

[43859:0x102801000]    36699 ms: Mark-sweep 1391.7 (1532.2) -> 1391.7 (1532.2) MB, 340.9 / 0.0 ms  allocation failure GC in old space requested
[43859:0x102801000]    37042 ms: Mark-sweep 1391.7 (1532.2) -> 1391.7 (1488.7) MB, 342.2 / 0.0 ms  last resort 
[43859:0x102801000]    37363 ms: Mark-sweep 1391.7 (1488.7) -> 1391.7 (1477.2) MB, 321.5 / 0.0 ms  last resort 


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x9bb715a9891 <JS Object>
    0: builtin exit frame: concat(this=0x9121f6720a1 <JS Array[712704]>,0x9121f6720a1 <JS Array[712704]>)

    1: combine [/Users/andrew/Documents/lunr-bug-repro/node_modules/lunr/lunr.js:~2319] [pc=0x254525acdf8d](this=0x1ef10c469561 <JS Object>,otherMatchData=0x9121f672619 <JS Object>)
    2: query [/Users/andrew/Documents/lunr-bug-repro/node_modules/lunr/lunr.js:1859] [pc=0x254525abd0e2](this...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::Factory::NewUninitializedFixedArray(int) [/usr/local/bin/node]
 5: v8::internal::Factory::NewJSArrayStorage(v8::internal::Handle<v8::internal::JSArray>, int, int, v8::internal::ArrayStorageAllocationMode) [/usr/local/bin/node]
 6: v8::internal::Factory::NewJSArray(v8::internal::ElementsKind, int, int, v8::internal::ArrayStorageAllocationMode, v8::internal::PretenureFlag) [/usr/local/bin/node]
 7: v8::internal::ElementsAccessor::Concat(v8::internal::Isolate*, v8::internal::Arguments*, unsigned int, unsigned int) [/usr/local/bin/node]
 8: v8::internal::(anonymous namespace)::Fast_ArrayConcat(v8::internal::Isolate*, v8::internal::BuiltinArguments*) [/usr/local/bin/node]
 9: v8::internal::Builtin_Impl_ArrayConcat(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/usr/local/bin/node]
10: 0x25452570415d
Abort trap: 6
```

**Comment by olivernn** [comment]
Ah right, got it now. Thanks for putting together the reproduction, it'll be a big help in coming up with a better way combining the matched metadata.

**Comment by olivernn** [comment]
Ok, I think there is a reasonably simple fix for this. The metadata held in the inverted index is [always an array](https://github.com/olivernn/lunr.js/blob/master/lib/builder.js#L171-L175), this makes cloning the data, to prevent mutation, more straightforward.

While I'm here though I want to take a closer look into the `lunr.MatchData` object, I have the feeling it can be tidied up a bit.

**Comment by olivernn** [comment]
I've just released a fix for this issue in version 2.1.2, let me know if it solves the issues you were seeing.

**Comment by i80and** [comment]
It does fix the issue!

However, I do have a concern. Take this for what it is, and obviously feel free to just consider this acceptable loss, but v2.1.2 comes with a hefty performance drop due to the deep copies. Specifically, I get the following performance numbers for a simple test of 1,000 queries:

Version | Seconds
--------|---------
https://github.com/olivernn/lunr.js#v2.1.2 | 19.1s
https://github.com/i80and/lunr.js/tree/fix-mutation | 14.1s

However, I can't come up with a test case where one yields different results from the other. This should be expected: MatchData.combine() never mutates an array. It only ever mutates the metadata object. This means that a shallow copy of the metadata object should be OK unless I'm missing something.

**Comment by olivernn** [comment]
Sigh, yeah, I had a feeling there would be a bit of a hit in performance. I think in general there is some time to be saved around the whole metadata and match data stuff.

The problem with using `Object.assign` is purely one of compatibility, Lunr currently works anywhere that supports ES5. For later releases (3.x) that might be relaxed, but at the moment I think its still a reasonable baseline.

I'm going to close this issue, but I am taking a look at `lunr.MatchData` with an eye on making some performance improvements.
